workshop/simple-sql.md

```diff
@@ -0,0 +1,211 @@
+* ptree_list_tuple([pid_hash, process_name]): a list of tuples that include the pid_hash and process name
+    * Example: ```[{'pid_hash': D608DE1B627B24B99FD3877CB7BD478C, 'process_name': cmd.exe}, {'pid_hash': AE8D851F89DE572F205A0B15C34E5E4E, 'process_name': explorer.exe}, ...```
+
+
+### What are the roots of processes? They *should* all be ntoskrnl.exe
+```sql
+SELECT
+    -- Slice the list to exclude the child process
+    ptree_list_tuples[-1:][1]['process_name'] AS path_root,
+    count(DISTINCT ptree_list_tuples[-1:][1]['pid_hash']) AS uniq_root_pid_hash,
+    median(level) AS median_path_depth,
+    max(level) AS max_path_depth,
+    count(*) AS num_process
+FROM
+    process_path
+GROUP BY ALL
+ORDER BY ALL
+;
+```
+
+### Paths relative to the executable, or, how many different ways is an executable spawned?
+```sql
+SELECT
+    p.process_name,
+    p.ptree_list_tuples[2:2][1]['process_name'] AS parent,
+    p.ptree_list_tuples[3:3][1]['process_name'] AS grand_parent,
+    p.ptree_list_tuples[4:4][1]['process_name'] AS great_grand_parent,
+    max(p.level) AS max_depth,
+    count(DISTINCT p.hostname) AS uniq_hosts,
+    count(*) AS num_executions
+FROM process_path AS p
+GROUP BY ALL
+ORDER BY p.process_name, parent, grand_parent, great_grand_parent
+;
+```
+
+### Processes using a certain file/port/ip
+
+_Ouch!_ This dataset has no file activity!
+```sql
+SELECT
+    process_name,
+    protocol,
+    remote_ip_addr,
+    remote_port,
+    count(*)
+FROM process_net_conn
+WHERE remote_port = 22
+GROUP BY ALL
+ORDER BY ALL
+;
+```
+
+### Potentially interesting inter-host communication
+Let's look for network connections between hosts that we have instrumented. the goal is to look for interesting cases where we can confirm the processes responsible for each end of the connection.
+
+In the Wintap data model, we create a key for network 5-tuples "conn_id" (a hash of local ip/port -> remote ip/port, protocol). In this definition, local/remote are relative to the host recording the data. As an example, that means for a given connection, this would be the data:
+
+```sql
+SELECT
+    hostname,
+    process_name,
+    protocol,
+    local_ip_addr,
+    local_port,
+    remote_ip_addr,
+    remote_port
+FROM process_net_conn
+WHERE conn_id = 'F6A1A412FE8988B22F0CD1295B7117C0'
+;
+```
+
+|Hostname|process_name|protocol|local_ip_addr|local_port|remote_ip_addr|remote_port|
+|--------|------------|--------|-------------|----------|--------------|-----------|
+|ACME-HH-HGC|ssh.exe|TCP|172.31.37.19|62435|172.31.34.133|22|
+|ACME-DC1|sshd.exe|TCP|172.31.34.133|22|172.31.37.19|62435|
+
+So now the goal is to join those 2 rows into a single row representing process->network->process. And that gets done with:
+```sql
+SELECT
+    l.hostname,
+    l.process_name,
+    l.protocol,
+    l.local_ip_addr,
+    l.remote_ip_addr,
+    r.hostname,
```

Comment:
we have `hostname` and `process_name` 2x in this select statement

Reply:
THey're duplicated on purpose as they represent the local (l) and remote(r) end of a process to process connection. I updated the example SQL and added a more complete explanation of what's going on.